### PR TITLE
Test run of lambda-on-binaris

### PR DESCRIPTION
**Pull request by rylandg**
Michael I just wanted to know your thoughts. I understand that this has huge implications but it seems like it might be worth the effort.

Note that  responses are not fully supported. We would need to write a simple wrapper that extracts the APIGateway response fields and then uses our native HTTPResponse.

### CODE REVIEW

File: lambda-on-binaris/package.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    "type": "git",
+    "url": "git+https://github.com/binaris/functions-examples.git"
+  },
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "bugs": {
+    "url": "https://github.com/binaris/functions-examples/issues"
+  },
+  "homepage": "https://github.com/binaris/functions-examples",
```
Comment (michaeladda) [inline lambda-on-binaris/package.json]: prefer direct page link

File: README.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ Each example contains a README.md with an explanation about the solution and it'
 | Example | Runtime  |
 |:--------------------------- |:-----|
 | [`fractal-terrain-generation`](fractal-terrain-generation) <br/> Render terrain in a browser with a backbend implemented in Serverless functions | nodeJS |
+| [`lambda-on-binaris`](lambda-on-binaris) <br/> Run simple AWS Lambda functions while gaining the benefits of the Binaris platform | nodeJS |
```
Comment (michaeladda) [inline README.md]: style: maybe lambda-wrapper ? (lambda on binaris is a bit weird. maybe it's just me)

File: lambda-on-binaris/.gitignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# package directories
```
Comment (michaeladda) [inline lambda-on-binaris/.gitignore]: these should probably in the top level directory of the repo, not?

File: lambda-on-binaris/package.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "lambda-on-binaris",
+  "version": "1.0.0",
+  "private": true,
```
Comment (michaeladda) [inline lambda-on-binaris/package.json]: maybe make it a real wrapper, and the example (=handler) can be its test
(not really sure if it's better, since i don't think this is really useful)

File: lambda-on-binaris/binaris.d.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/// <reference types="node" />
```
Comment (michaeladda) [inline lambda-on-binaris/binaris.d.ts]: why is this here?
isn't this a duplicate of https://github.com/binaris/binaris/blob/master/lib/functionTemplates/js/binaris.d.ts ?

oh, it's part of the new template. didn't occur to me this can daunting for javascript people. we should instead from the binaris package instead of importing type locally from a file with `@type { import("./binaris").BinarisFunction }`. this is not nice.

File: lambda-on-binaris/function.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+const { join } = require('path');
+const { execute } = require('lambda-local');
+
+function makeAPIGatewayEvent(body, context) {
+  const isBase64Encoded = (!body || body === {}) ? true : false;
```
Comment (michaeladda) [inline lambda-on-binaris/function.js]: i'm almost certain `!body` is a bug.
for example if `body === false`
moreso, isBase64Encoded is probably not correct either since 1. nothing about the encoding is suggested here and 2. in lambda the body is never parsed

File: lambda-on-binaris/function.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  return {
+    path: context.request.path,
+    httpMethod: context.request.method,
+    headers: context.request.headers,
```
Comment (michaeladda) [inline lambda-on-binaris/function.js]: did you verify single value here?

File: lambda-on-binaris/function.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    multiValueHeaders: context.request.headers,
+    queryStringParameters: context.request.query,
```
Comment (michaeladda) [inline lambda-on-binaris/function.js]: did you verify single value here?

File: lambda-on-binaris/function.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    multiValueQueryStringParameters: context.request.query,
+    pathParmeters: undefined,
```
Comment (michaeladda) [inline lambda-on-binaris/function.js]: you can use context.request.path

**Comment by rylandg** [comment]
@michaeladda 

Thanks for all the feedback. Clearly I will address all of the code concerns before/if we move forward but I realized that a PR might not be the right medium for what I'm trying to achieve.

There are many holes in the above code, for example we don't handle different response types or any of the associated complexities. What I want to know is whether 

1) You think this is a valuable API for us to support
2) Whether it's worth spending more time to refine and harden this

I realize now that a spec might have been a better medium but I think the above decisions are heavily reliant on the expected end user experience (code in this PR). Do you think a spec is a much better medium?
